### PR TITLE
Static row chart

### PR DESCRIPTION
**Pull request by alxnddr**
Epic https://github.com/metabase/metabase/issues/24877
Extracted independent and safely mergeable pieces from https://github.com/metabase/metabase/pull/25408

## Changes

Adds a static row chart. Commits are logically separated so I hope it will make the review process a bit easier considering the PR size.

### Tech details

Code that is used for static rendering cannot import any of the following from the main app code since it is highly coupled and brings dependencies that do not work in the static rendering:
- formatting 
- measuring text
- chart settings computation
- emotionJS

Due to these limitations, I structured components in the following way so that they reuse as much shared code as possible.

#### Environment agnostic
- `frontend/src/metabase/visualizations/components/RowChart/RowChartView` — dumb declarative view component that does not even calculate positions.
- `frontend/src/metabase/visualizations/components/RowChart` — component that calculates elements positions, margins, ticks so that they don't overlap (to be implemented), etc. It should not know anything about data structure, mbql, columns, drills logic, etc.

#### Environment sensitive
- `frontend/src/metabase/static-viz/components/RowChart` — static row chart root which provides static rendering specific formatters, text measurer, etc 
- Dynamic row chart is not in this PR but you can have a glimpse here https://github.com/metabase/metabase/pull/25408

## Screenshot

Screen Shot 2022-10-07 at 3 44 39 AM

## How to verify
- Rebuild the static viz bundle if you haven't, with this command
   ```bash
    yarn build-static-viz
    ```
- Create a row chart
- Add it to a dashboard
- Send the static version using dashboard subscriptions

#### Known issues that will be addressed in the following PRs
- Incorrectly formats date Y-axis
- Ticks and data labels are cut off near the right edge of a chart
- Lack of unit tests

### CODE REVIEW

**Review by ranquild — COMMENTED** [review]
General comments:
- I think we should make it explicit which code from `visualizations` is intended to be used in multiple environments. We could create `visualizations/core` or `visualizations/shared` folder  and put shared components there.
- I think we shound't pretend types for this chart are generic. I don't like the naming and placement of certain types, e.g. `ChartTheme`. I think they should be local to this chart type, and naming should make it clear, e.g. `RowChartTheme` or something. Same for "shared" utils that compute something for this chart.
- I think we should validate that the API of this shared component (e.g. `RowChart`) makes sense on its own. Could we create a storybook for it? I suppose we might need to improve the API to make it work.
- WDYT about following more strict structure like we used in `material` based projects before? e.g. not nesting component folders (multiple component files in the same folder is file).

File: src/metabase/pulse/render/body.clj
Comment (dpsutton) [inline src/metabase/pulse/render/body.clj:698]: this extra line should be removed

Comment (alxnddr) [reply src/metabase/pulse/render/body.clj]: @dpsutton done

**Comment by alxnddr** [comment]
@dpsutton I narrowed down the `card` to just `viz-settings` and the `data` has the same shape one returned by API handler for the main app dynamic chart. This is intentional and allows to have an isomorphic chart that can be rendered in both environments instead of having two separate implementations as we do for other static visualizations

File: frontend/src/metabase/visualizations/shared/hooks/use-two-dimensional-chart-series.ts
Comment (ranquild) [inline frontend/src/metabase/visualizations/shared/hooks/use-two-dimensional-chart-series.ts:9]: Should it be just a utility method? It doesn't seem that memoizing individual functions instead of the whole one gives any benefits. Or does it?

**Review by ranquild — APPROVED** [review]
The happy path works good. Still, there are some small differences here and there, e.g. in the number of ticks in each axis, which would be great to tackle in next PRs.

File: frontend/src/metabase/static-viz/components/RowChart/theme.ts
Comment (WiNloSt) [inline frontend/src/metabase/static-viz/components/RowChart/theme.ts:4]: I like how you define everything style related to the static chart in a single file that's so easy to manage, I should have done this for the gauge chart too, as now the values are spread across different files afaik.

**Review by WiNloSt — COMMENTED** [review]
I can only review a few files so far, this will take me like at least a few days to fully review this, but I will keep reviewing this every day.

File: frontend/src/metabase/lib/measure-text.ts
Comment (alxnddr) [reply frontend/src/metabase/lib/measure-text.ts:9]: It was used by the scalar visualization: https://github.com/metabase/metabase/pull/25829/files/2ecc6a25660550156baf9157739e2e0e4577f757#diff-675462a4272eee1a87d3a511c516d42bebd9223ff7e9e0620604f06436237046

File: frontend/src/metabase-types/api/dataset.ts
Comment (alxnddr) [reply frontend/src/metabase-types/api/dataset.ts:5]: `[[null]]` is a possible dataset value:
Screen Shot 2022-10-10 at 6 40 09 PM

File: frontend/src/metabase/visualizations/shared/hooks/use-two-dimensional-chart-series.ts
Comment (alxnddr) [reply frontend/src/metabase/visualizations/shared/hooks/use-two-dimensional-chart-series.ts]: The setting is called `graph.series_order` and it is sort of a single order of multiple series so I'm not sure about this change. Overall, I think it makes sense to stick with the spelling of the settings key. How does this sound?

File: frontend/src/metabase/visualizations/shared/utils/data.ts
Comment (alxnddr) [reply frontend/src/metabase/visualizations/shared/utils/data.ts:218]: Fixed, thanks!

File: frontend/src/metabase/static-viz/components/RowChart/RowChart.tsx
Comment (alxnddr) [reply frontend/src/metabase/static-viz/components/RowChart/RowChart.tsx]: Missed this one, thanks for pointing out

**Review by dpsutton — APPROVED** [review]

**Comment by dpsutton** [comment]
I'm approving and I suspect @adam-james-v will have some follow-up tests?

File: frontend/src/metabase/visualizations/shared/hooks/use-two-dimensional-chart-series.ts
Comment (WiNloSt) [reply frontend/src/metabase/visualizations/shared/hooks/use-two-dimensional-chart-series.ts]: Cool, didn't realize that. Makes sense then.

File: jest.unit.conf.json
Comment (WiNloSt) [inline jest.unit.conf.json:7]: Why do we need this settings for d3 libraries? Won't it compile without this setting?

**Review by WiNloSt — APPROVED** [review]
This is the 2nd attempt in reviewing this PR. I think it should be fairly safe to merge as the simple scenario works. I haven't tested this PR extensively, but so far I didn't find anything major that would prevent this PR from shipping.

I'll review this PR one more time tomorrow though, but feel free to merge if this unblock you for other stuff.

File: frontend/src/metabase/visualizations/lib/settings/stacking.ts
Comment (alxnddr) [reply frontend/src/metabase/visualizations/lib/settings/stacking.ts:3]: The reason I did that is that `stackable.stack_type` is specific to our codebase visual setting and on the `RowChart` level I operate with [d3 terms](https://github.com/d3/d3-3.x-api-reference/blob/master/Stack-Layout.md#offset). What do you think?

Comment (WiNloSt) [reply frontend/src/metabase/visualizations/lib/settings/stacking.ts:3]: That's fair given there's the root for these terms not just random ones. Though in d3 it's `stackOffset` not `stackingOffset` right?

File: jest.unit.conf.json
Comment (alxnddr) [reply jest.unit.conf.json:7]: Exactly, tests won't compile

File: frontend/src/metabase/static-viz/components/RowChart/utils/labels.ts
Comment (alxnddr) [reply frontend/src/metabase/static-viz/components/RowChart/utils/labels.ts:22]: This is pretty ugly but the logic behind is that when you define x-axis label for Line/Area/Bar chart and switch to the row chart, you probably expect to have it just rotated and see the label on the y-axis. The same when you switch from row to bar. In the current settings code there is no easy way to catch that viz type change to invert settings there so I do it here for now

File: frontend/src/metabase/static-viz/components/RowChart/utils/labels.ts
Comment (alxnddr) [reply frontend/src/metabase/static-viz/components/RowChart/utils/labels.ts:14]: I will try to address this in the main app row chart follow-up. The problem is that I'm unable to simply import the settings code with its default in the static environment, so I'll explore refactoring there

File: frontend/src/metabase/static-viz/components/RowChart/utils/labels.ts
Comment (WiNloSt) [reply frontend/src/metabase/static-viz/components/RowChart/utils/labels.ts:22]: Oh right. Yeah, when changing from Line/Area/Bar to Row it's just like rotating the chart 90degress so makes sense. 👍

File: frontend/src/metabase/static-viz/components/RowChart/utils/format.ts
Comment (alxnddr) [reply frontend/src/metabase/static-viz/components/RowChart/utils/format.ts:24]: Oh, thanks for pointing out, I'll fix that

File: frontend/src/metabase/static-viz/containers/StaticChart/constants.ts
Comment (alxnddr) [reply frontend/src/metabase/static-viz/containers/StaticChart/constants.ts:37]: Fixed, thanks!

File: frontend/src/metabase/static-viz/containers/StaticChart/StaticChart.tsx
Comment (alxnddr) [reply frontend/src/metabase/static-viz/containers/StaticChart/StaticChart.tsx]: Fixed 😄

**Review by WiNloSt — COMMENTED** [review]
this is the 3rd review. I think I just need another review to fully review this PR despite it being merged.

File: frontend/src/metabase/static-viz/components/RowChart/constants.ts
Comment (alxnddr) [reply frontend/src/metabase/static-viz/components/RowChart/constants.ts:55]: Nice catch, I will clean it up

File: frontend/src/metabase/visualizations/shared/components/RowChartView/RowChartView.tsx
Comment (WiNloSt) [reply frontend/src/metabase/visualizations/shared/components/RowChartView/RowChartView.tsx:131]: emotion won't work with static chart?

File: frontend/src/metabase/visualizations/shared/utils/data.ts
Comment (alxnddr) [reply frontend/src/metabase/visualizations/shared/utils/data.ts:139]: Actually, it is completely valid to have data without breakouts so probably I should not have `breakout: {}` in the initial value since it will be just an empty object for charts without a breakout. What do you think?

Comment (WiNloSt) [reply frontend/src/metabase/visualizations/shared/utils/data.ts:139]: That sounds better since `breakout` is optional in `GroupDatum` so using `Required<GroupDatum>` is kind of weird in a way that sometimes we expect `breakout` to be there and sometimes we don't.

**Pull request merged**
